discovery/kubernetes/kubernetes.go

```diff
@@ -233,8 +234,13 @@ func New(l log.Logger, conf *SDConfig) (*Discovery, error) {
 		kcfg.BearerToken = token
 
 		if conf.BasicAuth != nil {
-			kcfg.Username = conf.BasicAuth.Username
-			kcfg.Password = string(conf.BasicAuth.Password)
+			kcfg.WrapTransport = func(rt http.RoundTripper) http.RoundTripper {
```

Comment:
Why are we doing this by hand rather than relying on config_util for all of this?

Reply:
There is no way to pass a `http.Client` to the Kubernetes client...

Reply:
How about using NewRoundTripperFromConfig?

Reply:
I've done that but I'm a bit worried that it will change some connection settings under the hood. At least we can run prombench to check that it doesn't introduce some regression.

Reply:
All the settings are in the transport, the client is a plain wrapper.

Reply:
Still the `client-go` code does some tuning which would be "lost" if we pass our own `Transport`.

https://github.com/prometheus/prometheus/blob/12708acd1516e0d1756006a54684ccaa76f1aaf2/vendor/k8s.io/client-go/transport/cache.go#L85-L98

https://github.com/prometheus/prometheus/blob/12708acd1516e0d1756006a54684ccaa76f1aaf2/vendor/k8s.io/apimachinery/pkg/util/net/http.go#L104-L117

Reply:
That doesn't seem to be doing much we're not already doing ourselves.

Reply:
`Dialer.Timeout`, `Dialier.KeepAlive` and `TLSHandshakeTimeout` values aren't set by `NewRoundTripperFromConfig`. Wouldn't it make sense to initialize them to sane defaults?

Reply:
We use these with contexts, rather than having individual timeouts for each part of a connection. I'd presume k8 does the same.

Reply:
I've double checked in the client-go code and done some tests with a local environment and indeed relying only on context should be safe.